rfcs/text/0000-matrix-service.md

```diff
@@ -0,0 +1,418 @@
+
+#### Format
+
+The API generates expression matrices in three formats: `loom`, `mtx` and `csv`. From the [loom](http://loompy.org/) documentation page, the format is "an efficient file format for large omics datasets", enabling the storage of large expression matrices with metadata and also supports many popular data processing programming languages. For these reasons, `loom` is the service's default and most performant format. The `mtx` format adheres to the [10x feature-barcode matrix specification](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/output/matrices) to enable compatibility with 10X methods from `scanpy` and `Seurat`. `csv` is supported as a raw text format for which performance may not be optimized.
+
+#### Feature
+
+Users are also able to specify the expression feature type, i.e. the matrix's axis label, as `gene` (default) or `transcript` names.
+
+#### Species (proposed)
+
+`species` is a potential fifth parameter that the endpoint may accept. Currently, `species` is not a top-level parameter and is rather specified within `filter` via the `cell_suspension.genus_species.ontology` and `cell_suspension.genus_species.ontology_label` filters. Since a distinct matrix is generated for each `species` requested, it may be a more intuitive user experience for it to be a top-level parameter. Additionally, it requires non-trivial logic for the Matrix Service to parse a complex `filter` object to determine which species are requested, i.e. which species to generate a matrix for. In order to simplify the user's mental model of the API as well as the service's logic complexity, `species` is proposed as a fifth request body parameter.
+
+Note: Currently, supported species are _Homo Sapiens_ and _Mus Musculus_.
+
+#### Response
+
+The response object returns a `request_id` that users can use to retrieve the status and results of a request by supplying it to the `GET` `/matrix` endpoint. A `status` and `message` field are also supplied to inform the user whether the request started successfully or not.
+
+*Note about multiple species:*
+If a request selects cells from multiple species, the Matrix Service will generate one matrix per species and thus return multiple `request_id`s. The convention here is for the human matrix to be reflected in `request_id` and all other species matrices to be reflected in a list of `request_id`s via `non_human_request_ids`.
+
```

Comment:
Do you have a usability testing plan for this? from a very naive standpoint that feels like a lot of additional work to get the data for non-human species? What was the rationale to not return an array of matrix files?

Reply:
As the API evolves, we should be explicit about testing new changes such as this one in our future User Studies.

This design choice was made for a couple reasons:

1. So as to not introduce a breaking API change that changing the type returned by `request_id`
2. The initial design assumed 1 output matrix per 1 request ID and is reflected in design decisions throughout the software. Returning multiple matrices per request ID would compromise the design and introduce, in our opinion, unnecessary software complexity (e.g. many other 1:1 mappings in the code would need to support 1:N). This in addition to the project being the _Human Cell Atlas_ supported this cleaner, simpler implementation.

With that being said, the consideration for the user's experience is paramount and should this design prove to be inadequate for users retrieving non-human data, the team acknowledges that the software should accommodate the required design.